console-framework-client/src/main/java/io/axoniq/console/framework/application/RSocketDomainEntityDataResponder.kt

```diff
@@ -0,0 +1,112 @@
+import org.axonframework.lifecycle.Phase
+import org.axonframework.serialization.Serializer
+import org.slf4j.LoggerFactory
+
+open class RSocketDomainEntityDataResponder(
+        private val domainEventStreamProvider: DomainEventStreamProvider,
+        private val registrar: RSocketHandlerRegistrar,
+        private val domainEventAccessMode: DomainEventAccessMode,
+        private val serializer: Serializer,
+) : Lifecycle {
+    private val logger = LoggerFactory.getLogger(this::class.java)
+
+    override fun registerLifecycleHandlers(registry: Lifecycle.LifecycleRegistry) {
+        registry.onStart(Phase.EXTERNAL_CONNECTIONS, this::start)
+    }
+
+    fun start() {
+        registrar.registerHandlerWithPayload(
+                Routes.Enity.DOMAIN_EVENTS,
+                DomainEventsQuery::class.java,
+                this::handleDomainEventsQuery
+        )
+        registrar.registerHandlerWithPayload(
+                Routes.Enity.ENTITY_STATE_AT_SEQUENCE,
+                EntityStateAtSequenceQuery::class.java,
+                this::handleEntityStateAtSequenceQuery
+        )
+    }
+
+    private fun handleDomainEventsQuery(query: DomainEventsQuery): DomainEventsResult {
+        logger.debug("Handling AxonIQ Console DOMAIN_EVENTS query for request [{}]", query)
+        logger.debug("Access mode is set on: {}.", domainEventAccessMode)
+
+        val startIndex = query.page * query.pageSize
+
+        // Determine whether to include payload based on the access mode
+        val includePayload = domainEventAccessMode == DomainEventAccessMode.FULL ||
+                domainEventAccessMode == DomainEventAccessMode.PREVIEW_PAYLOAD_ONLY
+
+        // Fetch domain events starting from the calculated sequence number (offset)
+        val events = domainEventStreamProvider
+                .getDomainEventStream(query.entityId, startIndex.toLong()) ?: emptyList()
+
+        // If the result is empty, return an empty result but still indicate position
+        if (events.isEmpty()) {
+            return DomainEventsResult(
+                    entityId = query.entityId,
+                    entityType = "",
+                    domainEvents = emptyList(),
+                    page = query.page,
+                    pageSize = query.pageSize,
+                    totalCount = startIndex.toLong() // Total so far is up to this offset
+            )
+        }
+
+        // Map the first 'pageSize' events into the response model
+        val pagedEvents = events
+                .asSequence()
+                .take(query.pageSize)
+                .map { event ->
+                    DomainEvent(
+                            sequenceNumber = event.sequenceNumber,
+                            timestamp = event.timestamp,
+                            payloadType = event.payloadType.toString(),
+                            payload = if (includePayload)
+                                serializer.serialize(event.payload, String::class.java).data
```

Comment:
With this extension function:

```
private fun String?.truncate(i: Int): String? {
        if(this == null || this.length <= i) {
            return this
        }
        return this.substring(0, i) + "... [truncated to $i characters]"
    }
```

We can do here:

```suggestion
                                serializer.serialize(event.payload, String::class.java).data.truncate(50*1024)
```